### PR TITLE
Send HUP progress messages to balena API

### PR DESCRIPTION
**Pull request by kb2ma**
Adds a `notify_hup_progress()` API call, which updates balena API on status of a host OS update. Uses this call to notify of failure and eventual success. Supports incorporation of a takeover based OS update into balena's OS update tools.

Added option `--report-hup-progress` to activate these notifications. This option is considered internal, to be used only by balena's tools.

Also added a second commit to move the api_calls module from beneath stage1 to beneath common. This new call is made from init and stage2.

### CODE REVIEW

**Comment by kb2ma** [comment]
Pushed update for use of  `cargo fmt` to resolve test failures.

File: src/init.rs [hunk 1]
```diff
@@ -351,6 +353,13 @@ pub fn init() -> ! {
 
         setup_stage2_init_fallback_log(&s2_config.fallback_log_filename);
     }
+    // Required to send HUP progress messages to balena API.
```
Comment (rahul-thakoor) [inline src/init.rs:356]: i would update this comment to mention that this is required for network access after `pivot_root` is called

this enables API calls other than reporting HUP progress

**Comment by rahul-thakoor** [comment]
Tested on Intel NUC
```
2024-12-10 04:33:11 INFO  No stage 2 log device requested
2024-12-10 04:33:11 DEBUG [takeover::stage1] Stage 2 config: 
---
log_dev: ~
log_level: DEBUG
fallback_log: true
fallback_log_filename: advanced.20241210_0832.log
fallback_log_dirname: fallback_log
flash_dev: /dev/nvme0n1
pretend: false
umount_parts:
  - dev_name: /dev/nvme0n1p3
    mountpoint: /mnt/sysroot/inactive
    fs_type: ext4
  - dev_name: /dev/nvme0n1p5
    mountpoint: /var/volatile/lib/docker
    fs_type: ext4
  - dev_name: /dev/nvme0n1p2
    mountpoint: /mnt/sysroot/active
    fs_type: ext4
  - dev_name: /dev/nvme0n1p1
    mountpoint: /mnt/boot
    fs_type: vfat
  - dev_name: /dev/nvme0n1p4
    mountpoint: /var/volatile/lib/systemd
    fs_type: ext4
work_dir: /mnt/data
image_path: /mnt/data/balena-cloud-generic-amd64-6.0.24.img.gz
config_path: /mnt/data/config.4y172HGO1M.json
backup_path: ~
device_type: "X68_64/Intel Nuc,,"
tty: /dev/pts/1
api_endpoint: "https://api.balena-cloud.com"
api_key: <hidden>
uuid: d3d77cf007cf2a6d1e9c92e6378beabe
report_hup_progress: true
````


```
2024-12-10 04:34:16 INFO  [takeover::stage2] Migration completed successfully
2024-12-10 04:34:16 DEBUG [reqwest::connect] starting new connection: https://api.balena-cloud.com/
2024-12-10 04:34:17 DEBUG [takeover::common::api_calls] HUP progress result = Response { url: Url { scheme: "https", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("api.balena-cloud.com")), port: None, path: "/v6/device(uuid='d3d77cf007cf2a6d1e9c92e6378beabe')", query: None, fragment: None }, status: 200, headers: {"date": "Tue, 10 Dec 2024 04:34:17 GMT", "content-type": "application/json; charset=utf-8", "content-length": "4", "connection": "keep-alive", "cache-control": "no-cache", "etag": "W/\"4-P9ln0Jp0jh8rJtb+Vi5xVaqH6d4\"", "vary": "Accept-Encoding", "cf-cache-status": "DYNAMIC", "strict-transport-security": "max-age=15552000", "server": "cloudflare", "cf-ray": "8efa85887f883876-MRU", "alt-svc": "h3=\":443\"; ma=86400"} }
2024-12-10 04:34:17 INFO  [takeover::stage2] HUP progress notification OK

```

image

**Review by rahul-thakoor — APPROVED** [review]
tested, working on Intel NUC
HUP Progress reported to balenaCloud

File: src/init.rs (hunk 1 above)
Comment (kb2ma) [reply src/init.rs:356]: Yes, network setup needs to be more generic. At the same time it is dependent on the source OS for the location of the certificates and the DNS configuration. It feels like a good place to use an interface, which also allows us to customize the error message. If it fails from balenaOS running from a HUP script, that's a more serious problem than when running from Raspberry Pi OS.

**Pull request merged**